architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+apiVersion: radapp.io/v1alpha3
+metadata:
+  name: env
+  namespace: radius-system
+spec:
+  template: |
+    extension radius
+
+    param k8sNamespace string
+
+    resource env 'Applications.Core/environments@2023-10-01-preview' = {
+      name: 'env'
+      properties: {
+        compute: {
+          kind: 'kubernetes'
+          namespace: k8sNamespace
+        }
+      }
+    }
+  parameters: |
+    using './env.bicep'
+
+    param k8sNamespace string = 'demo'
+```
+
+## Design
+
+### High Level Design
+<!--
+High level overview of the data flow and key components.
+
+Provide a high-level description, using diagrams as appropriate, and top-level
+explanations to convey the architectural/design overview. Don’t go into a lot
+of details yet but provide enough information about the relationship between
+these components and other components. Call out or highlight new components
+that are not part of this feature (dependencies). This diagram generally
+treats the components as black boxes. Provide a pointer to a more detailed
+design document, if one exists. 
+-->
+
+This design proposes the creation of a new Kubernetes controller (BicepDeployment controller) packaged within the `radius-controller` service.
+
+The BicepDeployment controller will be implemented as a Kubernetes controller that watches and reconciles BicepDeployment resources on the cluster. The BicepDeployment controller will adhere to the following control loop for each of the BicepDeployment resources on the cluster:
+
+1. When a BicepDeployment resource is created or updated on the cluster, the controller will compile the Bicep template and parameters into ARM JSON. If there are any errors during the compilation, the controller will update the `status.phrase` and the `status.description` of the BicepDeployment resource to indicate that the deployment has failed.
+
+2. Check if we have an "operation" in progress. If so, check it's status.
+    1. If the operation is still in progress, then queue another reconcile (polling).
+	  2. If the operation completed successfully then update the status and continue processing (happy-path).
+	  3. If the operation failed then update the status and continue processing (retry).
+3. If the ApplicationDeployment is being deleted then process deletion.
+	  1. This may require us to start a DELETE operation. After that we can continue polling.
+    2. If the ApplicationDeployment is not being deleted then process this as a creation or update.
+	  3. This may require us to start a PUT operation. After that we can continue polling.
+  
+1. The controller will perform some simple lookups to Radius to determine if the provided template can be deployed. It will make two checks:
+    - Check if the template contains an empty `param environment`. If it does, the controller will query the Radius API to get an environment ID (must have exactly one environment if left unspecified like this) and deploy the template using that environment.
```

Comment:
This is the kind of thing that the rad CLI does today and I figured it'd be easy enough to implement. let me know what y'all think

Reply:
where do you think we'll get the application/environment names + scope?